<a href="https://colab.research.google.com/github/fullbread9977/Seoul-Bike-data-analysis/blob/main/%EB%94%B0%EB%A6%89%EC%9D%B4_%EC%9D%BC%EB%B3%84_%EB%B6%84%EC%84%9D_1~5%EC%9B%94.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

서울 열린데이터 광장 데이터사용 [다운로드](https://data.seoul.go.kr/dataList/OA-15246/F/1/datasetView.do#)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


구글 드라이브 런타임     
- 연결하여 내 드라이브안에 저장되어있는 파일을 쓰기 위함

In [ ]:
import pandas as pd
import pandas_profiling
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import font_manager, rc
plt.rcParams['font.family'] = 'NanumGothic'

**필요한 라이브러리 설치**   
- pandas : 엑셀같은 표를 그리기위해 사용되는 패키지
- pandas_profiling : 데이터의 상태를 보기위해 사용되는 패키지
- numpy : 행렬 연산에 편의성을 제공하는 패키지 
- matplotlib.pyplot : 데이터 시각화, 여러가지 표를 그리기 위해 사용하는 패키지
- seaborn : 데이터분포 시각화를 위한 패키지
- matplotlib: 그래프의 rc에 한글 사용을 위한 패키지
- '남산고딕'이라는 한글 폰트를 사용


In [ ]:
df = pd.read_csv("/content/drive/MyDrive/seoulbiohub/공공자전거 이용정보(일별)_2020.01_05.csv", 
                 encoding="cp949")
df

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
0        2020-01-01    101     101. (구)합정동 주민센터  ...  0.00     0.0      29
1        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3240.0      52
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
...             ...    ...                  ...  ...   ...     ...     ...
3052548  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      15
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[3052553 rows x 11 columns]

공공데이터 1~5월 사용, 데이터 3,052,553 건 , 11 개의 컬럼

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3052553 entries, 0 to 3052552
Data columns (total 11 columns):
 #   Column   Dtype  
---  ------   -----  
 0   대여일자     object 
 1   대여소번호    int64  
 2   대여소      object 
 3   대여구분코드   object 
 4   성별       object 
 5   연령대코드    object 
 6   이용건수     int64  
 7   운동량      object 
 8   탄소량      object 
 9   이동거리(M)  float64
 10  이용시간(분)  int64  
dtypes: float64(1), int64(3), object(7)
memory usage: 256.2+ MB


데이터의 상태를 보기 위해 df.info()를 입력, 운동량과 탄소량이 문자열로 되어있다.

In [ ]:
def clean(x):
    list(x)
    return x.strip().replace("\n","").replace('\\N',"")

문자열의 이상치를 제거하기 위해 /n을 전부 공백으로 바꿔주기 위해 작성한 함수 clean(x)

In [ ]:
df

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
0        2020-01-01    101     101. (구)합정동 주민센터  ...  0.00     0.0      29
1        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3240.0      52
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
...             ...    ...                  ...  ...   ...     ...     ...
3052548  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      15
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[3052553 rows x 11 columns]

표를 읽어온다

In [ ]:
df['운동량']

0            0.00
1           80.54
2           72.07
3          157.56
4           39.99
            ...  
3052548      0.00
3052549      0.00
3052550      0.00
3052551      0.00
3052552      0.00
Name: 운동량, Length: 3052553, dtype: object

운동량의 데이터타입은 문자열로 되어있는 것을 볼 수 있다

In [ ]:
df['운동량'] = df['운동량'].apply(lambda x : clean(x) )

운동량의 이상치를 apply를 사용하여 /n을 공백으로 대체

In [ ]:
df['운동량'] = pd.to_numeric(df['운동량']) 

운동량을 숫자형으로 형 변환을 해준다

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3052553 entries, 0 to 3052552
Data columns (total 11 columns):
 #   Column   Dtype  
---  ------   -----  
 0   대여일자     object 
 1   대여소번호    int64  
 2   대여소      object 
 3   대여구분코드   object 
 4   성별       object 
 5   연령대코드    object 
 6   이용건수     int64  
 7   운동량      float64
 8   탄소량      object 
 9   이동거리(M)  float64
 10  이용시간(분)  int64  
dtypes: float64(2), int64(3), object(6)
memory usage: 256.2+ MB


운동량을 숫자형으로 형 변환확인

In [ ]:
df['탄소량'] = df['탄소량'].apply(lambda x : clean(x) )

탄소량도 똑같이 이상치 값을 공백으로 변환해준다

In [ ]:
df['탄소량'] = pd.to_numeric(df['탄소량']) 

탄소량도 똑같이 숫자형 데이터로 변환한다

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3052553 entries, 0 to 3052552
Data columns (total 11 columns):
 #   Column   Dtype  
---  ------   -----  
 0   대여일자     object 
 1   대여소번호    int64  
 2   대여소      object 
 3   대여구분코드   object 
 4   성별       object 
 5   연령대코드    object 
 6   이용건수     int64  
 7   운동량      float64
 8   탄소량      float64
 9   이동거리(M)  float64
 10  이용시간(분)  int64  
dtypes: float64(3), int64(3), object(5)
memory usage: 256.2+ MB


운동량,탄소량 전부 float 형태의 숫자형으로 변환된것을 확인

In [ ]:
df.isna().sum()

대여일자            0
대여소번호           0
대여소             0
대여구분코드          0
성별         325967
연령대코드           0
이용건수            0
운동량          9166
탄소량          9166
이동거리(M)         0
이용시간(분)         0
dtype: int64

na(공백 값) 값이 몇개나 있는지 계산한다 성별에는 325967개 운동량은 9166, 탄소량도 9166개로 확인됬다 

In [ ]:
df.dropna()
df = df.dropna()

데이터가 300만 건이므로 데이터에 비해 na 값이 미미한 수치로 판정,na값을 제거 하기로 한다. 

In [ ]:
df.isna().sum()

대여일자       0
대여소번호      0
대여소        0
대여구분코드     0
성별         0
연령대코드      0
이용건수       0
운동량        0
탄소량        0
이동거리(M)    0
이용시간(분)    0
dtype: int64

제거가 잘됬는지 본다

In [ ]:
df.head(30)

대여일자  대여소번호               대여소  ...    탄소량  이동거리(M) 이용시간(분)
0   2020-01-01    101  101. (구)합정동 주민센터  ...   0.00      0.0      29
1   2020-01-01    101  101. (구)합정동 주민센터  ...   0.75   3240.0      52
2   2020-01-01    101  101. (구)합정동 주민센터  ...   0.75   3250.0      21
3   2020-01-01    101  101. (구)합정동 주민센터  ...   1.59   6860.0      46
4   2020-01-01    101  101. (구)합정동 주민센터  ...   0.23    990.0       5
5   2020-01-01    101  101. (구)합정동 주민센터  ...   0.23    990.0       5
6   2020-01-01    101  101. (구)합정동 주민센터  ...   0.14    610.0       2
7   2020-01-01    102   102. 망원역 1번출구 앞  ...   0.35   1520.0      65
8   2020-01-01    102   102. 망원역 1번출구 앞  ...   1.91   8230.0      40
9   2020-01-01    102   102. 망원역 1번출구 앞  ...   1.38   5940.0      52
10  2020-01-01    102   102. 망원역 1번출구 앞  ...   2.22   9580.0     157
11  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.95   4080.0      75
12  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.00      0.0      10
13  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.31   1330.0       7
15  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.12    500.0       5
16  2020-01-01    102   102. 망원역 1번출구 앞  ...   1.18   5050.0      38
17  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.64   2750.0      11
18  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.63   2730.0      66
19  2020-01-01    102   102. 망원역 1번출구 앞  ...  15.03  64810.0      66
20  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.26   1110.0       9
21  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.22    950.0       5
22  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.54   2310.0      22
23  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.20    870.0       6
24  2020-01-01    103   103. 망원역 2번출구 앞  ...   1.95   8420.0      48
25  2020-01-01    103   103. 망원역 2번출구 앞  ...   1.29   5580.0      67
26  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.22    960.0       3
28  2020-01-01    103   103. 망원역 2번출구 앞  ...   2.06   8900.0      78
29  2020-01-01    103   103. 망원역 2번출구 앞  ...   1.17   5020.0     118
30  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.67   2920.0      18
31  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.43   1840.0      17

[30 rows x 11 columns]

제대로 됬는지 보기위해 데이터의 30개만 불러온다

In [ ]:
df['성별'] = df['성별'].apply(lambda x : clean(x) )

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


성별도 /n값을 공백으로 변환한다

In [ ]:
# 성별이 M 혹은 F인 것에 해당하는 행만 추출
df2 = df[(df['성별'] == 'M') | (df['성별'] == 'F')]   
df2

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
5        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
6        2020-01-01    101     101. (구)합정동 주민센터  ...  0.14   610.0       2
...             ...    ...                  ...  ...   ...     ...     ...
3052543  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0     113
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[1455636 rows x 11 columns]

성별이 남자 또는 여자로 확실히 표기된 것들을 가져와 df2 이라는 데이터프레임에 새로 담는다

In [ ]:
df2

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
5        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
6        2020-01-01    101     101. (구)합정동 주민센터  ...  0.14   610.0       2
...             ...    ...                  ...  ...   ...     ...     ...
3052543  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0     113
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[1455636 rows x 11 columns]

데이터 전처리가 끝난 데이터 프레임, 총 145만 건이 남았다

In [ ]:
df2 = pd.get_dummies(df2, columns= ['연령대코드', '성별'])

연령대코드와 성별의 경우 서열이 있다고 판단 되지않아 명목코드로 값이 존재하는 곳에 1, 존재하지않는 곳에 0 표시로 one-hot 인코딩으로 컬럼을 늘려 데이터를 변환시킨다.

In [ ]:
df2

대여일자  대여소번호                  대여소  ... 연령대코드_AGE_008  성별_F  성별_M
2        2020-01-01    101     101. (구)합정동 주민센터  ...             0     1     0
3        2020-01-01    101     101. (구)합정동 주민센터  ...             0     1     0
4        2020-01-01    101     101. (구)합정동 주민센터  ...             0     0     1
5        2020-01-01    101     101. (구)합정동 주민센터  ...             0     0     1
6        2020-01-01    101     101. (구)합정동 주민센터  ...             1     0     1
...             ...    ...                  ...  ...           ...   ...   ...
3052543  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1

[1455636 rows x 19 columns]

바뀐 데이터 프레임 확인

In [ ]:
val_name = list(df2['대여구분코드'].unique())
val_name

['정기', '일일(회원)', '단체', '일일(비회원)', 'BIL_021']

대여구분코드 컬럼에 무슨 값이 존재하는지 가져온다

In [ ]:
df2 = df2.replace(val_name,[1,0,0,0,0])

정기권과 나머지 다른 권을 구분하고 싶으므로 정기권은 1, 다른 값들은 0으로 값을 변환한다.

In [ ]:
df2.head(30)

대여일자  대여소번호               대여소  ...  연령대코드_AGE_008  성별_F  성별_M
2   2020-01-01    101  101. (구)합정동 주민센터  ...              0     1     0
3   2020-01-01    101  101. (구)합정동 주민센터  ...              0     1     0
4   2020-01-01    101  101. (구)합정동 주민센터  ...              0     0     1
5   2020-01-01    101  101. (구)합정동 주민센터  ...              0     0     1
6   2020-01-01    101  101. (구)합정동 주민센터  ...              1     0     1
8   2020-01-01    102   102. 망원역 1번출구 앞  ...              0     0     1
9   2020-01-01    102   102. 망원역 1번출구 앞  ...              0     0     1
15  2020-01-01    102   102. 망원역 1번출구 앞  ...              0     1     0
16  2020-01-01    102   102. 망원역 1번출구 앞  ...              0     1     0
17  2020-01-01    102   102. 망원역 1번출구 앞  ...              0     1     0
18  2020-01-01    102   102. 망원역 1번출구 앞  ...              0     0     1
19  2020-01-01    102   102. 망원역 1번출구 앞  ...              0     0     1
20  2020-01-01    102   102. 망원역 1번출구 앞  ...              0     0     1
21  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     0     1
22  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     0     1
28  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     1     0
29  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     1     0
30  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     1     0
31  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     0     1
32  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     0     1
33  2020-01-01    103   103. 망원역 2번출구 앞  ...              0     0     1
38  2020-01-01    104   104. 합정역 1번출구 앞  ...              0     1     0
39  2020-01-01    104   104. 합정역 1번출구 앞  ...              0     0     1
45  2020-01-01    104   104. 합정역 1번출구 앞  ...              0     1     0
46  2020-01-01    104   104. 합정역 1번출구 앞  ...              0     0     1
47  2020-01-01    104   104. 합정역 1번출구 앞  ...              0     0     1
48  2020-01-01    104   104. 합정역 1번출구 앞  ...              0     0     1
49  2020-01-01    104   104. 합정역 1번출구 앞  ...              0     0     1
51  2020-01-01    105   105. 합정역 5번출구 앞  ...              0     1     0
52  2020-01-01    105   105. 합정역 5번출구 앞  ...              0     0     1

[30 rows x 19 columns]

제대로 확인했는 지 위해 데이터프레임 30개만 확인

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1455636 entries, 2 to 3052552
Data columns (total 19 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   대여일자           1455636 non-null  object 
 1   대여소번호          1455636 non-null  int64  
 2   대여소            1455636 non-null  object 
 3   대여구분코드         1455636 non-null  int64  
 4   이용건수           1455636 non-null  int64  
 5   운동량            1455636 non-null  float64
 6   탄소량            1455636 non-null  float64
 7   이동거리(M)        1455636 non-null  float64
 8   이용시간(분)        1455636 non-null  int64  
 9   연령대코드_AGE_001  1455636 non-null  uint8  
 10  연령대코드_AGE_002  1455636 non-null  uint8  
 11  연령대코드_AGE_003  1455636 non-null  uint8  
 12  연령대코드_AGE_004  1455636 non-null  uint8  
 13  연령대코드_AGE_005  1455636 non-null  uint8  
 14  연령대코드_AGE_006  1455636 non-null  uint8  
 15  연령대코드_AGE_007  1455636 non-null  uint8  
 16  연령대코드_AGE_008  1455636 non-null  uint8  
 17  성별_F    

df2의 데이터 형태가 어떻게 존재하는 지 확인하기 위해 작성

In [ ]:
colname = list(df2.columns)
colname

['대여일자',
 '대여소번호',
 '대여소',
 '대여구분코드',
 '이용건수',
 '운동량',
 '탄소량',
 '이동거리(M)',
 '이용시간(분)',
 '연령대코드_AGE_001',
 '연령대코드_AGE_002',
 '연령대코드_AGE_003',
 '연령대코드_AGE_004',
 '연령대코드_AGE_005',
 '연령대코드_AGE_006',
 '연령대코드_AGE_007',
 '연령대코드_AGE_008',
 '성별_F',
 '성별_M']

df2의 컬럼을 제대로 보기위해 컬럼을 불러오는 코드 작성

In [ ]:
df3 = df2[['대여구분코드',
 '이용건수',
 '운동량',
 '탄소량',
 '이동거리(M)',
 '이용시간(분)',
 '연령대코드_AGE_001',
 '연령대코드_AGE_002',
 '연령대코드_AGE_003',
 '연령대코드_AGE_004',
 '연령대코드_AGE_005',
 '연령대코드_AGE_006',
 '연령대코드_AGE_007',
 '연령대코드_AGE_008',
 '성별_F',
 '성별_M']]

df3에 새로 컬럼을 담는다

In [ ]:
df3.head()

대여구분코드  이용건수     운동량   탄소량  ...  연령대코드_AGE_007  연령대코드_AGE_008  성별_F  성별_M
2       1     1   72.07  0.75  ...              0              0     1     0
3       1     1  157.56  1.59  ...              0              0     1     0
4       1     1   39.99  0.23  ...              0              0     0     1
5       1     1   25.48  0.23  ...              0              0     0     1
6       1     1   17.63  0.14  ...              0              1     0     1

[5 rows x 16 columns]

제대로 담겨졌는지 df3의 상위 5개 데이터값만 가져온다



In [ ]:
df3.to_csv("bycycle_daily.csv", encoding="utf-8")

df3을 csv 파일 utf-8 형식으로 저장한다
- 파일 경로를 지정하지 않을 시 drive의 상위폴더인 content에 파일이 저장된다.


In [ ]:
from sklearn.model_selection import train_test_split

데이터들을 학습과 시험으로 구분하는 라이브러리

In [ ]:
X = df3.drop('대여구분코드', axis=1)
y = df3['대여구분코드']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

X는 대여구분코드를 제외한 모든 코드를 담고   
y는 대여구분코드만 담는다.   
데이터를 학습 데이터를 8, 시험데이터를 2로 나누어 담는다 랜덤방법을 100으로 섞어준다

# 로지스틱 회귀분석을 통한 정기권 구매 예측

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

로지스틱회귀분석 라이브러리를 불러온다    
매트릭스 라이브러리 정확도와 오차행렬을 불러온다

In [ ]:
model = LogisticRegression()

로지스틱회귀모델을 불러온다

In [ ]:
pip install h5py==2.10.0 

     |████████████████████████████████| 2.9 MB 4.2 MB/s 
  Attempting uninstall: h5py
    Found existing installation: h5py 3.1.0
    Uninstalling h5py-3.1.0:
      Successfully uninstalled h5py-3.1.0


라이브 러리 중 h5py을 2.10.0으로 재설치하여 다운그레이드한다

**에러 메세지**    
오류: pip의 종속성 해석기는 현재 설치된 모든 패키지를 고려하지 않습니다. 이 동작은 다음 종속성 충돌의 원인입니다.
tensorflow 2.6.0에는 h5py~=3.1.0이 필요하지만 호환되지 않는 h5py 2.10.0이 있습니다.

In [ ]:
model.fit(X_train, y_train)

LogisticRegression()

AttributeError: 'str' object has no attribute 'decode 의 경우   
h5py가 3 버전일 때 발생하는 에러이므로   
바로 위에 model = LogisticRegression() 을 누르고 +코드를 눌러 코드 칸을 추가한다.   
pip install h5py==2.10.0 입력하여 다운그레이드 해준다.

In [ ]:
pred = model.predict(X_test)

학습이 잘됬는지 검증할 데이터 X_test를 불러온다

In [ ]:
accuracy_score(y_test, pred)

0.7545478277596109

실제 시험해야되는 데이터 y_test를 불러 예측한다    
정확도는 75.45%

In [ ]:
confusion_matrix(y_test, pred)

array([[ 10177,  64119],
       [  7339, 209493]])

로지스틱의 회귀율은    
예측O 실제O 10177,    
예측X 실제O 64119,    
예측O 실제X 7339,    
예측X 실제X 209493 으로 나왔다

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(y_test, pred, digits=5))

              precision    recall  f1-score   support

           0    0.58101   0.13698   0.22169     74296
           1    0.76566   0.96615   0.85430    216832

    accuracy                        0.75455    291128
   macro avg    0.67333   0.55157   0.53800    291128
weighted avg    0.71854   0.75455   0.69286    291128



- 정밀도 76.56%
- 재현율 96.61%





# 디시즌 트리를 통한 이용권 구매 예측

In [ ]:
from sklearn.tree import DecisionTreeClassifier

디시즌 트리를 통한 이용권 구매예측 라이브러리를 불러온다

In [ ]:
model =  DecisionTreeClassifier(max_depth =10)

디시즌모델을 불러온다

In [ ]:
model.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=10)

아까 위에서 나눈 학습데이터와 시험데이터를 불러와 적용시킨다

In [ ]:
pred = model.predict(X_test)

제대로 훈련을 됬는 지 보기 위해 훈련을 시험해야할 X_test를 넣는다


In [ ]:
y_test

1018537    1
2473591    1
2187189    0
1351449    0
1584534    1
          ..
2084582    0
2460769    0
520236     1
349052     1
2613147    1
Name: 대여구분코드, Length: 291128, dtype: int64

데이터 예측 결과

In [ ]:
print(y_test, pred)

1018537    1
2473591    1
2187189    0
1351449    0
1584534    1
          ..
2084582    0
2460769    0
520236     1
349052     1
2613147    1
Name: 대여구분코드, Length: 291128, dtype: int64 [1 1 0 ... 1 1 1]


시험결과와 답을 비교한다..

In [ ]:
accuracy_score(y_test, pred)

0.7883954823994944

디시즌 트리의 정확도는 정확도 78.83%으로 측정되었다

# 트리의 깊이를 조절하며 최적값 탐색 depth = 12일 때 값이 가장 높음

In [ ]:
for i in range(2, 20):
    model = DecisionTreeClassifier(max_depth = i)
    model.fit(X_train, y_train)
    pred = model.predict(X_test)
    print(i, round(accuracy_score(y_test, pred), 4))

2 0.7448
3 0.7658
4 0.7672
5 0.775
6 0.7798
7 0.7836
8 0.7858
9 0.788
10 0.7884
11 0.7892
12 0.7892
13 0.7884
14 0.7865
15 0.7845
16 0.7818
17 0.7798
18 0.7761
19 0.7726


가지를 늘리며 최적화 된 값이 12번째 라는 걸 알게됬다

In [ ]:
model =  DecisionTreeClassifier(max_depth =12)  # depth= 12로 다시 분석

모델을 12로 조정하여 다시 분석한다

In [ ]:
model.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=12)

모델에 테스트 데이터를 불러와 학습시킨다

In [ ]:
pred = model.predict(X_test)

제대로 훈련을 됬는 지 보기 위해 훈련을 시험해야할 X_test를 넣는다


In [ ]:
accuracy_score(y_test, pred)

0.7891752081558627

정확도 78.91

In [ ]:
confusion_matrix(y_test, pred)

array([[ 30704,  43592],
       [ 17785, 199047]])

디시즌 트리는        
예측O 실제O 30693,    
예측X 실제O 43603,    
예측O 실제X 17773,    
예측X 실제X 199059 으로 나왔다

In [ ]:
print(classification_report(y_test, pred, digits=5))

              precision    recall  f1-score   support

           0    0.63322   0.41327   0.50013     74296
           1    0.82034   0.91798   0.86642    216832

    accuracy                        0.78918    291128
   macro avg    0.72678   0.66562   0.68327    291128
weighted avg    0.77259   0.78918   0.77294    291128



- 정밀도:82.03%
- 재현율:91.80%

# 랜덤포레스트

In [ ]:
from sklearn.ensemble import RandomForestClassifier

랜덤 포레스트 라이브러리를 불러온다

In [ ]:
model = RandomForestClassifier(max_depth = 10, random_state = 100)

모델을 불러온다. 모델의 깊이는 10, 데이터를 분류하는 방식은 100번 방법으로 사용한다

In [ ]:
model.fit(X_train, y_train)

RandomForestClassifier(max_depth=10, random_state=100)

데이터를 가져와 학습한다

In [ ]:
pred = model.predict(X_test)

제대로 훈련을 됬는 지 보기 위해 훈련을 시험해야할 X_test를 넣는다


In [ ]:
y_test

1018537    1
2473591    1
2187189    0
1351449    0
1584534    1
          ..
2084582    0
2460769    0
520236     1
349052     1
2613147    1
Name: 대여구분코드, Length: 291128, dtype: int64

정답을 비교할 y_test를 불러온다

In [ ]:
accuracy_score(y_test, pred)

0.7815703058448518

정확도는 78.15%

In [ ]:
confusion_matrix(y_test, pred)

array([[ 20344,  53952],
       [  9639, 207193]])

디시즌 트리는        
예측O 실제O 20344,    
예측X 실제O 53952,    
예측O 실제X 9639,    
예측X 실제X 207193 으로 나왔다

In [ ]:
print(classification_report(y_test, pred, digits=5))

              precision    recall  f1-score   support

           0    0.67852   0.27382   0.39018     74296
           1    0.79340   0.95555   0.86696    216832

    accuracy                        0.78157    291128
   macro avg    0.73596   0.61468   0.62857    291128
weighted avg    0.76408   0.78157   0.74529    291128



- 정밀도는 79.34%
- 재현율은 95.55%

# 최적 max_depth 찾기

랜덤포레스트의 최적값을 찾기 위해 반복문으로 max_depth의 값을 늘려가며 모델을 학습을 시킨다. 학습용 데이터를 불러와서 모델을 학습시키고, 제대로 됬는지 시험용 데이터와 정답을 불러와 정확도를 본다.   
i는 횟수, 횟수의 정확도

In [ ]:
for i in range(2,20):
    model = RandomForestClassifier(max_depth=i, random_state=100)
    model.fit(X_train, y_train)
    pred = model.predict(X_test)
    print(i, accuracy_score(y_test, pred))

2 0.7447995383473935
3 0.7447995383473935
4 0.7451292902091177
5 0.7496977274600863
6 0.7583537138303427
7 0.7678134703635514
8 0.7706094913577533
9 0.7758580418235278
10 0.7815703058448518
11 0.78488843395345
12 0.7877016295237833
13 0.7894328267978347
14 0.7902778159435025
15 0.7909819735648924
16 0.7912979857657113
17 0.791411337968179
18 0.7910747162760023
19 0.790988843395345


17에서 79.14%의 정확도로 가장높게 나온다

In [ ]:
model = RandomForestClassifier(max_depth=17, random_state=100)
model.fit(X_train, y_train)
pred = model.predict(X_test)
print( accuracy_score(y_test, pred))

0.791411337968179


17번째로 값을 수정하고 코드실행

In [ ]:
from sklearn.metrics import log_loss

교차 엔트로피 손실 라이브러리를 불러온다

In [ ]:
logloss = log_loss(y_test, pred)

y_test 오차율의 정확도를 구한다

In [ ]:
confusion_matrix(y_test, pred)

array([[ 29927,  44369],
       [ 16357, 200475]])

랜덤포레스트는        
예측O 실제O 29927,    
예측X 실제O 44369,    
예측O 실제X 16357,    
예측X 실제X 200475 으로 나왔다


In [ ]:
accuracy_score(y_test, pred)

0.791411337968179


정확도 79.14%

In [ ]:
from sklearn.metrics import classification_report

분류 성능 평가 라이브러리를 불러온다

In [ ]:
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.65      0.40      0.50     74296
           1       0.82      0.92      0.87    216832

    accuracy                           0.79    291128
   macro avg       0.73      0.66      0.68    291128
weighted avg       0.77      0.79      0.77    291128



최적화 값의 정확도를 본다.     
정확도가 반올림되어 나오기때문에 좀 더 정확히 보기에 숫자가 5자리까지 나오도록 수정한다.

In [ ]:
print(classification_report(y_test, pred, digits=5))

              precision    recall  f1-score   support

           0    0.64659   0.40281   0.49638     74296
           1    0.81879   0.92456   0.86847    216832

    accuracy                        0.79141    291128
   macro avg    0.73269   0.66369   0.68243    291128
weighted avg    0.77484   0.79141   0.77351    291128



- 정밀도: 81.87%
- 재현율: 92.45%